### PR TITLE
Add purl getter to ArtifactCoordinates

### PR DESCRIPTION
**Pull request by blaumeiser-at-bosch**
Add the support for purls : https://github.com/package-url/purl-spec to artifacts. This is done as getter in the ArtifactCoordinates fact.

Will be tested within the policy engine, as soon as this PR is through

### CODE REVIEW

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java [hunk 1]
```diff
@@ -37,6 +37,11 @@ public DotNetCoordinates(String packageId, String version) {
     @Override
     public String getVersion() { return version; }
 
+    @Override
+    public String getPurl() {
+        return String.format("pkg:nuget/%s@%s", getName(), getVersion());
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java]: In general, purl-components need to be percent-encoded, see

https://github.com/heremaps/oss-review-toolkit/blob/673c9b817cb71fed1bcd313d51d4b7788fbd2e61/model/src/main/kotlin/Identifier.kt#L126
https://github.com/package-url/purl-spec#rules-for-each-purl-component

File: modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/Artifact.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package org.eclipse.sw360.antenna.drools;
```
Comment (maxhbr) [inline modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/Artifact.java]: missing license headers

File: modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/PolicyEngine.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package org.eclipse.sw360.antenna.drools;
```
Comment (maxhbr) [inline modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/PolicyEngine.java]: missing license headers

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java [hunk 4]
```diff
@@ -62,6 +62,11 @@ public String getVersion() {
         return getBundleVersion();
     }
 
+    @Override
+    public String getPurl() {
+        return String.format("pkg:p2/%s@%s", getName(), getVersion());
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java]: Instead of having this per coordinates implementation, can't we implement his in the base class and instead have the coordinates classes only implement a `getType()` method that is called by the base class implementation? At the example of this class, `getType()` should return "p2".

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java (hunk 1 above)
Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java]: Also, all implementations except the Maven one seem to miss the "namespace" part.

File: modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/Artifact.java (hunk 2 above)
Comment (maxhbr) [reply modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/Artifact.java]: are these changes related to this PR?

File: modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/PolicyEngine.java (hunk 3 above)
Comment (maxhbr) [reply modules/drools/drools-checker/src/main/java/org/eclipse/sw360/antenna/drools/PolicyEngine.java]: are these changes related to this PR?

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java (hunk 4 above)
Comment (maxhbr) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java]: just `getType` might not be enough: e.g. Docker has multiple ways of building a purl (with tag, with hash, ...)

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java [hunk 5]
```diff
@@ -14,6 +14,7 @@
 public abstract class ArtifactCoordinates<T extends ArtifactCoordinates> implements ArtifactIdentifier<T> {
     public abstract String getName();
     public abstract String getVersion();
+    public abstract String getPurl();
```
Comment (maxhbr) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: How do we plan to handle coordinates with insufficient information (e.g. with missing version or missing groupId)?

Comment (maxhbr) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: How do we plan to handle coordinates types, which are not covered by the current purl definition? Throwing `UnsupportedOperationException`? Maybe add this to the interface?

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java (hunk 4 above)
Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java]: Right, still the `type` would be "docker" in all cases. `qualifiers` and `subpath` components could be added later without breaking the existing code.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java (hunk 5 above)
Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: > How do we plan to handle coordinates types, which are not covered by the current purl definition?

Do you have any concrete example for such a type? Ideally we'd fix this by contributing to [purl-spec](https://github.com/package-url/purl-spec). But as an interim solution, yes, probably throw an exception.

Comment (maxhbr) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: I think for example stackage is not supported. (Unimportant: I am even unsure how something like stackage would be supported, where the version is attached to stackage itself and references consistent package sets, instead of individual versions for every single component.)

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java [hunk 6]
```diff
@@ -11,10 +11,29 @@
...
+    public abstract String getType();
+
+    public String getPurl() {
+        return String.format("pkg:%s/%s@%s", percentEncoding(getType()), percentEncoding(getName()),
+                percentEncoding(getVersion()));
+    }
+
+    protected String percentEncoding(String input) {
+        try {
+            return URLEncoder.encode(input, "UTF-8");
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: Just `encode` is unfortunately not enough, encoding "+" needs to be handled explicitly, see https://github.com/heremaps/oss-review-toolkit/blob/673c9b817cb71fed1bcd313d51d4b7788fbd2e61/utils/src/main/kotlin/Extensions.kt#L260-L263

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java (hunk 5 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: I do not think that missing information is an issue. In this case the component information is not sufficient, i.e., the purl is not valid and cannot be used. In this case, the artifact is not identified, so the compliance information has a major issue and therefore is only preliminary. The user has to solve that anyway.

The first usage of the purl is in the policy engine to make a unique hash of a policy violation, so that we can identify violations over different evaluations. I would prefer that the method always returns a String, even if the data is not complete. An unidentified artifact in the policy engine is a blocking violation, so the uniqueness of the violation issue is not that important, since it has to be solved.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java (hunk 1 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java]: I added this in my latest update. Concerning the namespace part, I think this is a minor bug in ORT, if I compare the examples from https://github.com/package-url/purl-spec:

pkg:npm/%40angular/animation@12.3.1
pkg:npm/foobar@12.3.1

pkg:nuget/EnterpriseLibrary.Common@6.0.1304

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java [hunk 7]
```diff
@@ -11,10 +11,29 @@
...
+
 public abstract class ArtifactCoordinates<T extends ArtifactCoordinates> implements ArtifactIdentifier<T> {
     public abstract String getName();
+
     public abstract String getVersion();
 
+    public abstract String getType();
+
+    public String getPurl() {
+        return String.format("pkg:%s/%s@%s", percentEncoding(getType()), percentEncoding(getName()),
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: The "namespace" component is still missing from here. In general a purl looks like: `pkg:<type>/<namespace>/<name>/<version>?<qualifiers>#<subpath>`, but the `?<qualifiers>#<subpath>` could be omitted for now.

If a package manager does not support a namespace, an empty string should be used, resulting in two consecutive `/` in the purl.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java (hunk 4 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/BundleCoordinates.java]: I changed that as you @sschuberth proposed. All coordinates we have now, are sufficiently encoded with the current functionality. So if we face more difficult Purl cases, the idea is to overwrite the getPurl method. Nonetheless, I like the idea that you can get a repository type from the coordinates.

**Comment by blaumeiser-at-bosch** [comment]
@sschuberth @maxhbr Changes are made as requested, resp. I commented on some remarks

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java [hunk 8]
```diff
@@ -61,6 +61,17 @@ public String getVersion() {
         return version;
     }
 
+    @Override
+    public String getType() {
+        return "maven";
+    }
+
+    @Override
+    public String getPurl() {
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: This specialization is not necessary with the generalization of namespaces, see above.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java (hunk 1 above)
Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/dotnet/DotNetCoordinates.java]: Oh, you're right, "The name is prefixed by a '/' separator when the namespace is not empty"!

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java (hunk 7 above)
Comment (maxhbr) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: but in the purl spec there are also examples like `pkg:gem/jruby-launcher@1.1.2?platform=java`

Comment (maxhbr) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: The purl spec even explicitly says, that this segment is not allowed to be empty:

> - **namespace**:
>
>  - The optional `namespace` contains zero or more segments, separated by slash   '/'
>  - Leading and trailing slashes '/' are not significant and should be stripped    in the canonical form. They are not part of the `namespace`
>  - Each `namespace` segment must be a percent-encoded string
>  - When percent-decoded, a segment:
>
>    - must not contain a '/'
>    - **must not be empty**
>
>  - A URL host or Authority must NOT be used as a `namespace`. Use instead a    `repository_url` qualifier. Note however that for some types, the    `namespace` may look like a host.

Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: See discussion above, that is correct

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java (hunk 8 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: I think, here we are in a more general discussion. If I add a getNamespace() method to ArtifactCoordinates, the whole concept of Artifact Coordinates could be simplified to one class as you did in ORT. But I did not want to open this box of Pandora for now. Making this overwrite is simplifying the world for all other coordinate types we have so far.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java (hunk 6 above)
Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: How was this resolved? Did you forget to push?

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java [hunk 9]
```diff
@@ -11,10 +11,36 @@
...
 public abstract class ArtifactCoordinates<T extends ArtifactCoordinates> implements ArtifactIdentifier<T> {
     public abstract String getName();
+
     public abstract String getVersion();
 
+    public abstract String getType();
+
+    public String getPurl() {
+        return String.format("pkg:%s/%s%s", percentEncoding(getType()), percentEncoding(getName()),
+                percentEncoding(getVersionForPurl()));
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: This also encodes the "@" separator for the version. I suggest to move `percentEncoding()` to inside `getVersionForPurl()` at line 31 instead.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java [hunk 10]
```diff
@@ -11,10 +11,36 @@
...
+
 public abstract class ArtifactCoordinates<T extends ArtifactCoordinates> implements ArtifactIdentifier<T> {
     public abstract String getName();
+
     public abstract String getVersion();
 
+    public abstract String getType();
+
+    public String getPurl() {
+        return String.format("pkg:%s/%s%s", percentEncoding(getType()), percentEncoding(getName()),
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: The `type` must not be percent-encoded, see https://github.com/package-url/purl-spec#rules-for-each-purl-component.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java [hunk 11]
```diff
@@ -61,6 +61,17 @@ public String getVersion() {
...
+        return String.format("pkg:%s/%s/%s%s", percentEncoding(getType()), percentEncoding(getGroupId()),
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: The `type` must not be percent-encoded, see https://github.com/package-url/purl-spec#rules-for-each-purl-component.

**Comment by sschuberth** [comment]
Just FYI, there's also https://github.com/package-url/packageurl-java which could be taken into use. We're not doing so for ORT yet either, but I'm considering it.

**Comment by maxhbr** [comment]
Alternatively to the comment https://github.com/eclipse/antenna/pull/267#issuecomment-524819103 by @sschuberth, I would suggest to encapsulate the purl stuff in something like:
```java
package org.eclipse.sw360.antenna.model.util;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;
import java.util.Optional;

public class PurlUtils {
    static private String percentEncoding(String input) {
        try {
            return URLEncoder.encode(input, "UTF-8").replace("+", "%20");
        } catch (UnsupportedEncodingException e) {
            throw new IllegalStateException("Programming error: Unknown encoding", e);
        }
    }

    static public String buildPurl(String type, List<String> path, Optional<String> version) {
        final StringBuilder stringBuilder = new StringBuilder("pkg:");
        stringBuilder.append(type);
        path.stream()
                .map(PurlUtils::percentEncoding)
                .forEach(s -> {
                        stringBuilder.append("/");
                        stringBuilder.append(s);
                });
        version.map(PurlUtils::percentEncoding)
                .ifPresent(v -> {
                stringBuilder.append("@");
                stringBuilder.append(v);
        });
        return stringBuilder.toString();
    }
}
```

**Comment by blaumeiser-at-bosch** [comment]
I will switch to the PURL library, that seems to be closer to the spec source.

**Comment by sschuberth** [comment]
Can we please squash the commits in this PR into a single commit now? Otherwise reviewing commit by commit is really misleading now, as the last commit voids it all.

**Comment by blaumeiser-at-bosch** [comment]
> Can we please squash the commits in this PR into a single commit now? Otherwise reviewing commit by commit is really misleading now, as the last commit voids it all.

@sschuberth: Done

**Comment by blaumeiser-at-bosch** [comment]
Azure is green: https://dev.azure.com/EclipseSW360antenna/EclipseSW360antenna/_build/results?buildId=414

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java [hunk 12]
```diff
@@ -70,7 +81,7 @@ public boolean isEmpty() {
 
     @Override
     public boolean matches(ArtifactIdentifier artifactIdentifier) {
-        if(artifactIdentifier instanceof MavenCoordinates) {
+        if (artifactIdentifier instanceof MavenCoordinates) {
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: Nit: Strictly speaking, the unrelated whitespace changes here and below should be moved out to a separate commit.

Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: Unfortunately a result of reformatting the file after my changes above this. Really removing that from the file?

Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: Well, I leave it to how strictly you handle this (hence only a "nit"). In ORT, we don't allow to mix topics within a single commit, and that includes formatting, as we try to separate functional from non-functional changes for more speedy reviews.

Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: FWIW, I didn't mean to *drop* the change from this PR, but just to separate it out into another commit in *this* PR.

Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/java/MavenCoordinates.java]: These changes were not intended at all, they were result of auto formatting, so I discarded them. In general, perhaps we should define at some point a common formatting and apply this over time to the source base. But this is not part of this change :-)

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java [hunk 13]
```diff
@@ -11,10 +11,36 @@
...
+    protected PackageURLBuilder addPurlFacts(PackageURLBuilder builder) {
+        return builder;
+    }
+
+    public PackageURL getPurl() {
+        PackageURLBuilder builder =  PackageURLBuilder.aPackageURL()
+                .withType(getType())
+                .withName(getName());
+
+        if (getVersion() != null) {
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: This explicit `null` check is still there, did you intentionally keep it?

Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: Version is according to the purl spec an optional value. I thought that it makes sense to not set the value, if the version is unknown instead of setting a null value, so yes this is intentional.

**Comment by blaumeiser-at-bosch** [comment]
Latest Azure build green: https://dev.azure.com/EclipseSW360antenna/EclipseSW360antenna/_build/results?buildId=420

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java (hunk 13 above)
Comment (sschuberth) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: My point is: Exactly that logic is already handled by the purl library (see the code I had linked in my original comment), so why also handle that case here?

Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactCoordinates.java]: Ok accepted, actually, there was a major bug in the implementation, but now it is fixed

**Comment by blaumeiser-at-bosch** [comment]
Azure is green: https://dev.azure.com/EclipseSW360antenna/EclipseSW360antenna/_build/results?buildId=425, therefore merging, although Travis not run.
